### PR TITLE
[webui][api] Refactor destroying Project/Package objects

### PR DESCRIPTION
**Pull request by hennevogel**

### CODE REVIEW

**Comment by adrianschroeter** [comment]
Are the commit_opts not a problem with parallel running requests when both modifying same project/package?

**Comment by hennevogel** [comment]
No as we are only using them in memory

**Comment by hennevogel** [comment]
Okay updated with your suggestions

**Comment by bgeuken** [comment]
The package deletion doesn't happen in a transaction anymore, right? This might lead to inconsistent states, eg. when the backend call results in some error.

**Comment by hennevogel** [comment]
All destroy calls are in a transaction. When writing to the backend goes wrong, destroy fails. That's why the sequence of before_destroy callbacks (including `dependent: destroy` on associations) is crucial.

File: src/api/app/controllers/webui/project_controller.rb [hunk 1]
```diff
@@ -250,7 +250,6 @@ def destroy
     if @project.can_be_really_deleted?
       parent = @project.parent
       Project.transaction do
-        @project.delete_on_backend
         @project.destroy
```
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: The transaction isn't needed any more, right?

File: src/api/app/models/bs_request_action.rb [hunk 2]
```diff
@@ -480,11 +467,15 @@ def source_cleanup_delete_path
...
 
+      source_project.commit_opts[:comment] = self.bs_request.description
+      source_project.commit_opts[:request] = self.bs_request.id
       source_project.destroy
       return "/source/#{self.source_project}"
     end
     # just remove one package
     source_package = source_project.packages.find_by_name!(self.source_package)
+    source_package.commit_opts[:comment] = self.bs_request.description
+    source_package.commit_opts[:request] = self.bs_request.id
```
Comment (bgeuken) [inline src/api/app/models/bs_request_action.rb]: What is this needed for?

**Comment by bgeuken** [comment]
The patchinfo related part looks good to me.

I would have expected more test changes and new tests given the amount of changes. Also it would be nice to have some more information in the commit message about the changes that have been made and the why.

File: src/api/app/controllers/webui/project_controller.rb [hunk 3]
```diff
@@ -250,7 +250,6 @@ def destroy
     if @project.can_be_really_deleted?
```
Comment (bgeuken) [inline src/api/app/controllers/webui/project_controller.rb]: haha

File: src/api/app/models/project.rb [hunk 4]
```diff
@@ -135,16 +139,8 @@ def cleanup_before_destroy
...
-    #        will be removed during that
-#    revoke_requests
-  end
-
-  def delete_on_backend(comment = nil)
-    path = source_path
-    path << Suse::Backend.build_query_from_hash({user: User.current.login, comment: comment}, [:user, :comment])
-    Suse::Backend.delete path
+    # revoke all requests that have this project as source/target
+    revoke_requests
```
Comment (bgeuken) [inline src/api/app/models/project.rb]: The issue mentioned few lines above won't happen anymore?

File: src/api/app/controllers/webui/project_controller.rb (hunk 1 above)
Comment (hennevogel) [inline src/api/app/controllers/webui/project_controller.rb]: Right

File: src/api/app/models/bs_request_action.rb (hunk 2 above)
Comment (hennevogel) [inline src/api/app/models/bs_request_action.rb]: To pass around some hash specific to this object (in memory)

File: src/api/app/models/project.rb (hunk 4 above)
Comment (hennevogel) [inline src/api/app/models/project.rb]: Nope it won't

Comment (bgeuken) [inline src/api/app/models/project.rb]: Cool

Comment (adrianschroeter) [inline src/api/app/models/project.rb]: On Dienstag, 6. Oktober 2015, 08:34:39 CEST wrote Bj�rn Geuken:
> > @@ -135,16 +139,8 @@ def cleanup_before_destroy
> >        end
> >      end
> >  
> > -    # revoke all requests
> > -    # FIXME: this is breaking request accepting with cleanup, if the project
> > -    #        will be removed during that
> > -#    revoke_requests
> > -  end
> > -
> > -  def delete_on_backend(comment = nil)
> > -    path = source_path
> > -    path << Suse::Backend.build_query_from_hash({user: User.current.login, comment: comment}, [:user, :comment])
> > -    Suse::Backend.delete path
> > +    # revoke all requests that have this project as source/target
> > +    revoke_requests
> 
> Cool
> 

revoking when it is used as target sounds wrong, shouldn't it be declined then?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imend�rffer, Jane Smithard, Graham Norton, HRB 21284 (AG N�rnberg) 
 
Maxfeldstra�e 5                         
90409 N�rnberg 
Germany

Comment (hennevogel) [inline src/api/app/models/project.rb]: I've just moved this functionality here. It did the same when deleting via API previously....

**Comment by hennevogel** [comment]
Now that #1205 is merged, can we merge this too finally? @adrianschroeter

**Comment by adrianschroeter** [comment]
commit_opts are not only used in memory as far as I can see. Can we avoid the database change to be on the save side?

**Pull request merged**